quarkus.jdt.ext/com.redhat.microprofile.jdt.quarkus/src/main/java/com/redhat/microprofile/jdt/internal/quarkus/scheduler/SchedulerUtils.java

```diff
@@ -0,0 +1,113 @@
+*     Red Hat Inc. - initial API and implementation
+*******************************************************************************/
+package com.redhat.microprofile.jdt.internal.quarkus.scheduler;
+
+import java.time.Duration;
+import java.time.format.DateTimeParseException;
+import java.util.regex.Pattern;
+
+/**
+ * Utility class for quarkus scheduler
+ *
+ */
+public class SchedulerUtils {
+
+	// Day of week and day of month cannot coexist
+	private static String[] cronPatterns = { "^[1-5]?[0-9]$", "^[1-5]?[0-9]$", "^(2[0-3]|1[0-9]|[1-9])$",
+			"^(3[01]|[12][0-9]|[1-9])$", "^(1[0-2]|[1-9])$", "[*?a-zA-Z0-9]+", "^(19|20)[0-9]{2}$" };
+
+	private static String envPattern = "^\\$\\{[0-9a-zA-Z\\.]*\\}$";
+
+	private static String looseEnvPattern = "^.*\\$\\{.*\\}.*$";
+
+	private SchedulerUtils() {
+	}
+
+	/**
+	 * Validate the @Scheduled cron member with each cron string part and return an
+	 * error message if necessary
+	 *
+	 * @param cronString the cron member value
+	 * @return the error fault for the cron string validation and null if valid
+	 */
+	public static SchedulerErrorCodes validateCronPattern(String cronString) {
+
+		String[] cronParts = cronString.split("\\s+");
+
+		if (cronParts.length < 6 || cronParts.length > 7) {
+			return SchedulerErrorCodes.INVALID_CRON_LENGTH;
+		}
+
+		for (int i = 0; i < cronParts.length; i++) {
+			String patternString = String.format("(%s|\\*|\\?|^\\*\\/\\d+)", cronPatterns[i]);
+			if (!Pattern.matches(patternString, cronParts[i])) {
+				return SchedulerErrorCodes.valueOf(i);
+			}
+		}
+		return null;
+	}
+
+	/**
+	 * Validate the string from the @Scheduled member can be parsed to a Duration
+	 * unit
+	 *
+	 * @param timePattern the member value for the time pattern
+	 * @return the INVALID_DURATION_PARSE_PATTERN error code if invalid and null if
+	 *         valid
+	 */
+//	\\d+[smhSMH]
+	public static SchedulerErrorCodes validateDurationParse(String timePattern) {
+		if (Pattern.matches("\\d+[smhSMH]$", timePattern)) {
+			return null;
+		}
+		try {
+			Duration.parse(timePattern);
+			return null;
+		} catch (DateTimeParseException e) {
+			return SchedulerErrorCodes.INVALID_DURATION_PARSE_PATTERN;
+		}
+	}
+
+	/**
+	 * Check if the member value env variable pattern is well formed
+	 *
+	 * @param memberValue the member value from expression
+	 * @return INVALID_CHAR_IN_EXPRESSION if the env variable is malformed
+	 */
+	public static SchedulerErrorCodes matchEnvMember(String memberValue) {
+		return checkLooseEnvPattern(memberValue)
+				? (checkEnvPattern(memberValue) ? SchedulerErrorCodes.VALID_EXPRESSION : SchedulerErrorCodes.INVALID_CHAR_IN_EXPRESSION)
+				: null;
+	}
+
+	/**
+	 * Match the member to an env variable pattern
+	 *
+	 * @param memberValue the member value from expression
+	 * @return true if pattern is well-formed env variable, false otherwise
+	 */
+	private static boolean checkEnvPattern(String memberValue) {
+		return Pattern.matches(envPattern, memberValue);
```

Comment:
For the `Pattern.matches(..)` calls, you may want to consider defining the regular expression as :

`private static final Pattern MY_PATTERN = Pattern.compile(SOME_REGEX_PATTERN)` and using it in the form `MY_PATTERN.matcher(inputString).matches()`. The reason for this is that on every keystroke relating to Scheduled annotations, the regular expressions are re-compiled every single time so it's best to avoid it.